### DISCUSSION TITLE
Sign with Google Cloud and a service account key file

### DISCUSSION THREAD

**Discussion by obones** [ebourg/jsign#173] [Q&A] [answered]
Hello,

I'm experimenting jsign with Google Cloud KMS and it works just fine with the syntax recommended in the documentation.
However, this uses a token from the `gcloud` CLI and it proves a bit difficult to automate on a headless system for me.
There is, however, the possibility to give google CLI tools the location of a json file containing the secret key for a service account to impersonate when using the API.
It's done via setting an environment variable:

```
set GOOGLE_APPLICATION_CREDENTIALS=C:\Path\To\credentials.json                             
```

This works fine with the CNG provider from Google, but I can't use it everywhere just like I would with jsign.
How can I specify the same value to jsign? 
I tried removing the `storepass` option but it says it is required.
I tried giving a blank value (`--storepass ""`) but it failed fetching the private key.
I tried using the `file:` prefix to point to the json file, but it stopped saying `java.lang.IllegalArgumentException: Illegal character(s) in message header value`

Thanks a lot for any suggestion.

**Comment by ebourg** [comment, marked as answer]
Jsign doesn't support the `GOOGLE_APPLICATION_CREDENTIALS` variable, you have to use the gcloud CLI to turn the secret key from `credentials.json` into an access token.

**Reply by obones** [reply]
Well, I have done this series of Windows commands:

```dos
gcloud auth activate-service-account --key-file=C:\Path\To\credentials.json
gcloud auth print-access-token > token
java -jar jsign-5.0.jar --storetype GOOGLECLOUD --storepass "file:./token" --keystore projects/MyProject/locations/MyLocation/keyRings/MyKeyRing --alias MyKeyName --tsaurl http://mytsserver/rfc3161 --tsmode RFC3161 --tsretrywait 1 --alg SHA-256 --url https://www.mydomain.com --certfile MyCertificate.cer "C:\Path\To\Binary.exe"
```

And this yields the following result:

```
jsign: Failed to retrieve the private key from the keystore
java.security.UnrecoverableKeyException: Unable to fetch Google Cloud private key 'projects/MyProject/locations/MyLocation/keyRings/MyKeyRing/cryptoKeys/MyKeyName'
        at net.jsign.jca.GoogleCloudSigningService.getPrivateKey(GoogleCloudSigningService.java:146)
        at net.jsign.jca.SigningServiceKeyStore.engineGetKey(SigningServiceKeyStore.java:41)
        at java.security.KeyStore.getKey(KeyStore.java:1023)
        at net.jsign.SignerHelper.build(SignerHelper.java:333)
        at net.jsign.SignerHelper.sign(SignerHelper.java:388)
        at net.jsign.JsignCLI.execute(JsignCLI.java:132)
        at net.jsign.JsignCLI.main(JsignCLI.java:40)
Caused by: java.io.IOException: 403 - PERMISSION_DENIED: Permission 'cloudkms.cryptoKeyVersions.get' denied on resource 'projects/MyProject/locations/MyLocation/keyRings/MyKeyRing/cryptoKeys/MyKeyName' (or it may not exist).
        at net.jsign.jca.RESTClient.query(RESTClient.java:98)
        at net.jsign.jca.RESTClient.get(RESTClient.java:54)
        at net.jsign.jca.GoogleCloudSigningService.getPrivateKey(GoogleCloudSigningService.java:130)
        ... 6 more
Try `java -jar jsign.jar --help' for more information.
```

I tried by placing the token directly in the command line, replicating the commands under Linux to no avail.
I verified that the key exists at the indicated location. 
I then tried with `signtool` and the `Google CNG provider` with the same credentials and it worked.

I must have missed something with jsign, but I can't to figure out what.

**Reply by ebourg** [reply]
Could you try appending the version of the key to the alias:

    --alias MyKeyName:1

**Reply by ebourg** [reply]
Do you know if the account used as a full access to they key, or only to the `asymmetricSign` operation?

**Reply by obones** [reply]
I did not know about the colon syntax, but I forgot to mention that I tried with this:

    --alias MyKeyName/cryptoKeyVersions/1 

Neither syntax changed the situation.

I have, however, found the solution after having activated audit logs in the Google Cloud account.
The service account that I am using initially had only the "Cloud KMS CryptoKey Signer/Verifier" role because it's the only one required for the CNG provider. It allows those operations: GetPublicKey, AsymetricSign and Verify
Unfortunately, this is not enough for Jsign which uses ListCryptoKeyVersions or GetCryptoKeyVersion depending on the presence of the version.
Adding the "Cloud KMS Viewer" role gives the right for the service account to perform those two operations and fixes the issue I was having.

In the end there is no issue in Jsign itself, just that one should keep in mind that the required rights for it differ slightly from the ones recommended by google itself.

**Reply by ebourg** [reply]
Interesting, thank you. I'll add that to the documentation. I guess the CNG provider expects a fully qualified key name with the version included and doesn't need to call ListCryptoKeyVersions. Jsign doesn't require the version and tries to guess it automatically, the benefit is that the jsign settings don't have to be updated when a new version is deployed.

I was wrong with the colon syntax, the right syntax is actually:

    --alias MyKeyName/cryptoKeyVersions/1:RSA

Could you give it a try and confirm it works with the "CryptoKey Signer/Verifier" role only?

**Reply by obones** [reply]
Yes, the CNG provider indeed expects a fully qualified key version and fails with just the key name.

Where does the value after colon come from?
I have tried with `--alias MyKeyName/cryptoKeyVersions/1` and just the "CryptoKey Signer/Verifier" role but it fails because jsign tries to use `GetCryptoKeyVersion` which is not granted by this role. 
In the same situation, the CNG uses `GetPublicKey` just before calling `AsymmetricSign` but I don't know what it does with the result of the first call nor if it would be sufficient for jsign usage.

**Reply by ebourg** [reply]
Jsign needs to know the algorithm of the key (RSA or ECDSA), if it isn't provided it calls `GetCryptoKeyVersion` to get it. If the algorithm is appended after the version (for example `myKeyName/cryptoKeyVersions/1:RSA`), Jsign makes a direct call to `asymmetricSign` and doesn't require the viewer role.

**Reply by obones** [reply]
That's good to know, in my case it would ECDSA. 
One thing to note though, is that the `GetPublicKey` API returns the algorithm for the key, if I read the [documentation](https://cloud.google.com/kms/docs/reference/rest/v1/projects.locations.keyRings.cryptoKeys.cryptoKeyVersions/getPublicKey) correctly. So maybe Jsign could use this API if the algorithm isn't provided and still not require the viewer role.

**Reply by ebourg** [reply]
I've updated the documentation to clarify the syntax of the alias parameter and the permissions required, thank you for your input. Also, the algorithm is now fetched from the certificate associated to the key, so the colon syntax is no longer necessary and one less API call is performed.

**Reply by obones** [reply]
Thanks, I'll make sure to update my scripts when a new version is out.
